package/libgluonutil/src/libgluonutil.c

```diff
@@ -212,6 +214,42 @@ char * gluonutil_get_domain(void) {
 	return ret;
 }
 
+char * gluonutil_get_primary_domain(void) {
+	if (!gluonutil_has_domains())
+		return NULL;
+
+	char *domain_code = gluonutil_get_domain();
+	if (!domain_code)
+		return NULL;
+
+	const char *domain_path_fmt = "/lib/gluon/domains/%s.json";
+	char domain_path[strlen(domain_path_fmt) + strlen(domain_code)];
+	snprintf(domain_path, sizeof(domain_path), domain_path_fmt, domain_code);
+	free(domain_code);
+
+	char primary_domain_path[PATH_MAX+1];
+	char *primary_domain_code;
+	ssize_t len = readlink(domain_path, primary_domain_path, PATH_MAX);
+	if (len < 0) {
+		// EINVAL = file is not a symlink = the domain itself is the primary domain
+		if (errno != EINVAL)
+			return NULL;
+
+		primary_domain_code = basename(domain_path);
```

Comment:
Hmm, I guess we could just return `domain_code` in this branch after all if we move the `free` below this point.

The following lines could be moved out of the `else` branch then.